controllers/cloud.redhat.com/clowdapp_reconciliation.go

```diff
@@ -0,0 +1,381 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	crd "github.com/RedHatInsights/clowder/apis/cloud.redhat.com/v1alpha1"
+	"github.com/RedHatInsights/clowder/controllers/cloud.redhat.com/clowderconfig"
+	"github.com/RedHatInsights/clowder/controllers/cloud.redhat.com/config"
+	"github.com/RedHatInsights/clowder/controllers/cloud.redhat.com/errors"
+	"github.com/RedHatInsights/clowder/controllers/cloud.redhat.com/providers"
+	provutils "github.com/RedHatInsights/clowder/controllers/cloud.redhat.com/providers/utils"
+	rc "github.com/RedHatInsights/rhc-osdk-utils/resource_cache"
+	"github.com/go-logr/logr"
+	"github.com/prometheus/client_golang/prometheus"
+	core "k8s.io/api/core/v1"
+	k8serr "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type ClowdAppReconciliation struct {
+	cache                 *rc.ObjectCache
+	recorder              record.EventRecorder
+	ctx                   *context.Context
+	client                client.Client
+	log                   *logr.Logger
+	app                   *crd.ClowdApp
+	req                   *ctrl.Request
+	env                   *crd.ClowdEnvironment
+	reconciliationMetrics ReconciliationMetrics
+}
+
+func MakeClowdAppReconciliation(ctx context.Context,
+	cache *rc.ObjectCache,
+	client client.Client,
+	recorder record.EventRecorder,
+	app *crd.ClowdApp,
+	log logr.Logger,
+	req *ctrl.Request) ClowdAppReconciliation {
+	return ClowdAppReconciliation{
+		cache:    cache,
+		recorder: recorder,
+		ctx:      &ctx,
+		client:   client,
+		log:      &log,
+		app:      app,
+		req:      req,
+	}
+}
+
+func (r *ClowdAppReconciliation) steps() []func() (ctrl.Result, error) {
+	return []func() (ctrl.Result, error){
+		r.getApp,
+		r.setPresentAndManagedApps,
+		r.startMetrics,
+		r.isAppMarkedForDeletion,
+		r.addFinalizer,
+		r.isEnvLocked,
```

Comment:
I wonder if we could/should rename some of these a little and combine them. Right now there are a lot of steps to visually go through, for example, what about
`getClowdEnv` also chain calling, `isEnvCoreNamespaceDeleted`, `isClowdEnvReady`, and `isClowdEnvReconciled`

Perhaps it needs a different name, but I think that would collapse things down a little in this list of steps.

It could also be argued that the same could be done for `getApp`.

Reply:
I would like to push back on this suggestion. My goal was for each step to (as much as possible) do a single thing and to have a complexity score of 1 or 2 - 3 max. This makes each step easy to reason about, and will make it very easy to write tests for each step. If the steps do more then there will be high conditionality, more ways to return, and will make testing more complex. (Note: I know tests aren't included in the patch. I ran out of time. But in future I'd like to add tests for every step.)

I do think this is the best way to bake this particular cake.